### PR TITLE
ggml : add support for dynamic loading of backends

### PR DESCRIPTION
**Pull request by slaren**
Adds support for loading backends dynamically at load time, without needing to link to them in the build.

- Building the backends as dynamic libraries can be enabled by building with cmake with `GGML_BACKEND_DL` enabled
- Adds the function `ggml_backend_load(const char * path)` to load a backend dynamically
- Adds the convenience function `ggml_backend_load_all(void)` to load all the known backends
- Adds the function `ggml_backend_unload(ggml_backend_reg_t reg)` to unregister and unload a backend
- Adds the optional function `ggml_backend_get_features` to obtain a list of flags of a backend. This replaces the calls to the `ggml_cpu_has_xx` functions from the CPU backend in llama.cpp
- In addition to the CPU backend, the CUDA backend also implements `ggml_backend_get_features`, which returns the list of archs included in the build and the build flags used such as `GGML_CUDA_FORCE_MMQ`. Other backends should also implement this function to report compile-time flags and features.

#### TODO
- [x] Version checking to avoid loading incompatible backends
- [x] Fix `ggml_backend_load_all` search paths

### CODE REVIEW

File: ggml/src/ggml-backend-impl.h [hunk 1]
```diff
@@ -204,12 +204,30 @@ extern "C" {
...
+        #else
+        #    define GGML_BACKEND_DL_IMPL(reg_fn)                          \
+                GGML_BACKEND_API ggml_backend_reg_t ggml_backend_init();  \
+                ggml_backend_reg_t ggml_backend_init() {                  \
+                    return reg_fn();                                      \
+                }
+        #endif
+    #else
+    #    define GGML_BACKEND_DL_IMPL(reg_fn)
+    #endif
```
Comment (ggerganov) [inline ggml/src/ggml-backend-impl.h:230]: Fix compile warnings:

```
llama.cpp/ggml/src/ggml-metal/ggml-metal.m:4400:1: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
 4400 | GGML_BACKEND_DL_IMPL(ggml_backend_metal_reg)
      | ^
llama.cpp/ggml/src/ggml-metal/../ggml-backend-impl.h:223:70: note: expanded from macro 'GGML_BACKEND_DL_IMPL'
  223 |                 GGML_BACKEND_API ggml_backend_reg_t ggml_backend_init();  \
      |                                                                      ^
```

```suggestion
    // Add backend dynamic loading support to the backend
    #ifdef GGML_BACKEND_DL
        #ifdef __cplusplus
        #    define GGML_BACKEND_DL_IMPL(reg_fn)                                 \
                extern "C" {                                                     \
                    GGML_BACKEND_API ggml_backend_reg_t ggml_backend_init(void); \
                }                                                                \
                ggml_backend_reg_t ggml_backend_init(void) {                     \
                    return reg_fn();                                             \
                }
        #else
        #    define GGML_BACKEND_DL_IMPL(reg_fn)                             \
                GGML_BACKEND_API ggml_backend_reg_t ggml_backend_init(void); \
                ggml_backend_reg_t ggml_backend_init(void) {                 \
                    return reg_fn();                                         \
                }
        #endif
    #else
    #    define GGML_BACKEND_DL_IMPL(reg_fn)
    #endif
```

**Review by ggerganov — APPROVED** [review]

**Comment by MaggotHATE** [comment]
Is `ggml_backend_load_all()` supposed to be called in static builds too? If I don't use it, there's a noticeable reduction in quality of generated answers. When used, it tries to load any backend .dll it can find, which probably shouldn't happen on a static build (especially with a backend - OPENBLAS, for example). Am I doing something wrong?

**Comment by slaren** [comment]
It is ok to call `ggml_backend_load_all` even on static builds since it allows loading external backends. If it cannot find dynamically loadable backends present, it won't do anything. The reduction in quality that you are observing is not likely to be caused by this.

**Comment by MaggotHATE** [comment]
> it allows loading external backends.

It does, but allocates memory again, essentially duplicating total memory usage. I suppose it's a mistake on my end? It shouldn't behave like that on a combination of a static build with a dynamic backend?

**Comment by slaren** [comment]
I could happen if you have a static backend and the same backend as a dynamic backend, but that does not happen normally, because backends build without `GGML_BACKEND_DL` enabled cannot be loaded dynamically.

**Comment by paoletto** [comment]
is this exposed in llama-server?

**Pull request merged**